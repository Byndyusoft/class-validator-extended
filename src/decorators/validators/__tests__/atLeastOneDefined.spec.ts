/*
 * Copyright 2021 Byndyusoft
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { IsOptional, IsString, validate } from "class-validator";

import { AtLeastOneDefined } from "~/src";

function getUserClass(properties?: string[]): {
  new (): {
    email?: string;
    name?: string;
  };
} {
  class User {
    @IsString()
    @IsOptional()
    public readonly email?: string;

    @IsString()
    @IsOptional()
    public readonly name?: string;

    @AtLeastOneDefined(properties)
    protected readonly _atLeastOneDefined?: unknown;
  }

  return User;
}

describe("decorators/validators/AtLeastOneDefined", () => {
  it("must forbid empty user", async () => {
    const user = new (getUserClass())();

    const validationErrors = await validate(user);

    expect(validationErrors).toMatchSnapshot();
  });

  it("must forbid user if some properties not defined", async () => {
    const user = new (getUserClass(["email", "name"]))();

    const validationErrors = await validate(user);

    expect(validationErrors).toMatchSnapshot();
  });
});
