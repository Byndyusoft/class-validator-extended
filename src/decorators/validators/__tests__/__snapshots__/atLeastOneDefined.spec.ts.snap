// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorators/validators/AtLeastOneDefined must forbid empty user 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "atLeastOneDefined": "at least one property in User must be defined",
    },
    "property": "_atLeastOneDefined",
    "target": User {},
    "value": undefined,
  },
]
`;

exports[`decorators/validators/AtLeastOneDefined must forbid user if some properties not defined 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "atLeastOneDefined": "at least one property in User of 'email', 'name' must be defined",
    },
    "property": "_atLeastOneDefined",
    "target": User {},
    "value": undefined,
  },
]
`;
